### PR TITLE
[FEATURE] Adding functions for implementing shopping cart

### PR DESCRIPTION
**Pull request by turbo-ele**

### CODE REVIEW

File: Api/OrderAddress.php [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+    public function getId()
+    {
+        return $this->entity->getId();
+    }
+
+    /**
+     * Set salutation
+     *
+     * @param string $salutation
+     *
```
Comment (christianbader) [inline Api/OrderAddress.php]: ?

File: Cart/CartManager.php [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+     */
+    public function __construct(
+        ObjectManager $em,
+        SessionInterface $session,
+        OrderRepository $orderRepository,
+        OrderManager $orderManager,
+        GroupedItemsPriceCalculatorInterface $priceCalculation,
+        $defaultCurrency
+    )
+    {
```
Comment (christianbader) [inline Cart/CartManager.php]: (
    $param,
    $paran
) {
}

File: Cart/CartManager.php [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        $this->em = $em;
+        $this->session = $session;
+        $this->orderRepository = $orderRepository;
+        $this->orderManager = $orderManager; //FIXME: unused
```
Comment (christianbader) [inline Cart/CartManager.php]: remove

File: Cart/CartManager.php [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+     * @param null $locale
+     *
+     * @return null|\Sulu\Bundle\Sales\OrderBundle\Entity\Order
+     */
+    public function getUserCart($user = null, $locale = null, $currency = null)
+    {
+        // cart by session ID
+        if (!$user) {
+            // TODO: get correct locale
+            // TODO: QUESTION: add locale to order?
```
Comment (christianbader) [inline Cart/CartManager.php]: add issue

File: Cart/CartManager.php [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            $locale = 'de';
+            $cartArray = $this->findCartBySessionId();
+        } else {
+            // TODO: check if cart for this sessionId exists and assign it to user
```
Comment (christianbader) [inline Cart/CartManager.php:120]: remove

File: Cart/CartManager.php [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+     * removes all elements from database but the first
+     *
+     * @param $cartArray
+     */
+    private function cleanupCartArray(&$cartArray)
+    {
+        if ($cartArray && count($cartArray) > 0) {
+            // handle cartArray count is > 1
+            foreach ($cartArray as $index => $cart) {
+                // TODO: QUESTION: delete expired carts?
```
Comment (christianbader) [inline Cart/CartManager.php]: issue

File: Cart/CartManager.php [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+     * @param null $locale
+     *
+     * @return null|Order
+     * @throws ItemNotFoundException
+     */
+    public function updateItem($itemId, $data, $user = null, $locale = null)
+    {
+        $cart = $this->getUserCart($user, $locale);
+        $userId = $user ? $user->getId() : null;
+
```
Comment (christianbader) [inline Cart/CartManager.php:261]: remove

File: Entity/OrderRepository.php [hunk 8]
```diff
@@ -109,6 +115,32 @@ public function findByIdAndLocale($id, $locale)
...
+    {
+        try {
+            $qb = $this->getOrderQuery($locale)
+                ->andWhere('o.creator = :user')
+                ->setParameter('user', $user)
+                ->andWhere('status.id = :statusId')
+                ->setParameter('statusId', $statusId)
+                ->orderBy('o.created', 'DESC');
+            
+            // TODO use expiryDate
```
Comment (christianbader) [inline Entity/OrderRepository.php]: remove

File: Order/BaseOrderManager.php [hunk 9]
```diff
@@ -0,0 +1,518 @@
+<?php
```
Comment (christianbader) [inline Order/BaseOrderManager.php]: remove file

File: Order/OrderManager.php [hunk 10]
```diff
@@ -221,8 +233,8 @@ function ($contact) use ($order) {
         $order->setCustomerName($customerName);
 
         // set OrderAddress data
-        $this->setOrderAddress($order->getInvoiceAddress(), $data['invoiceAddress'], $contact, $account);
-        $this->setOrderAddress($order->getDeliveryAddress(), $data['deliveryAddress'], $contact, $account);
+        $this->setOrderAddress($order->getEntity()->getInvoiceAddress(), $data['invoiceAddress'], $contact, $account);
```
Comment (christianbader) [inline Order/OrderManager.php:236]: length?

File: Tests/Functional/Manager/CartManagerTest.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        // get cart by user
+        $cart = $this->getCartManager()->getUserCart($this->data->user);
+
+        // $cart is an ApiOrder, so get entity first
+        $this->assertEquals($cart->getEntity(), $this->data->order);
+    }
+    
+    public function testGetCartBySessionId()
+    {
+        // TODO: fix mock of session
```
Comment (christianbader) [inline Tests/Functional/Manager/CartManagerTest.php:43]: remove this test

File: Tests/OrderDataSetup.php [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+        $this->em = $entityManager;
+
+        $this->loadFixtures();
+        $this->setUpTestData();
+    }
+    
+    public function setupCartTests()
+    {
+        // set order to cart order
+        $this->orderStatus = $this->em->getRepository(static::$orderStatusEntityName)->find(OrderStatus::STATUS_IN_CART);
```
Comment (christianbader) [inline Tests/OrderDataSetup.php]: length?

File: Tests/OrderTestBase.php [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        $this->assertEqualsIfExists($address->getPostboxCity(), $orderAddress, 'postboxCity');
+        $this->assertEqualsIfExists($address->getPostboxPostcode(), $orderAddress, 'postboxPostcode');
+
+        // account
+        if ($account) {
+            $this->assertEqualsIfExists($account->getName(), $orderAddress, 'accountName');
+            $this->assertEqualsIfExists($account->getUid(), $orderAddress, 'uid');
+        }
+
+        // TODO: check phone
```
Comment (christianbader) [inline Tests/OrderTestBase.php]: remove

File: Order/OrderManager.php (hunk 10 above)
Comment (turbo-ele) [inline Order/OrderManager.php:236]: < 150

File: Cart/CartManager.php (hunk 7 above)
Comment (turbo-ele) [inline Cart/CartManager.php:261]: done

File: Tests/Functional/Manager/CartManagerTest.php (hunk 11 above)
Comment (turbo-ele) [inline Tests/Functional/Manager/CartManagerTest.php:43]: nope :P

**Comment by christianbader** [comment]
ack

**Pull request merged**